docs/proposals/002-refresh-certificates.md

```diff
@@ -0,0 +1,371 @@
+
+Due to the very disruptive nature of the operation, refreshing the cluster CA
+is kept intentionally out of scope for the existing proposal, and will be dealt
+with separately in the future.
+
+### External CA certificate
+
+When deploying Canonical Kubernetes with an external CA certificate, the cluster
+does not know the CA key. Therefore, it is unable to sign new node certificates.
+
+Due to this, it is the responsibility of the orchestration tool that is used to
+sign the new certificates and make them available on the cluster nodes.
+
+Automating any part of this manual process is out of scope for this proposal.
+
+# Implementation Details
+
+## API Changes
+<!--
+This section MUST mention any changes to the k8sd API, or any additional API
+endpoints (and messages) that are required for this proposal.
+
+Unless there is a particularly strong reason, it is preferable to add new v2/v3
+APIs endpoints instead of breaking the existing APIs, such that API clients are
+not affected.
+-->
+
+Certificate refreshes behave differently for control plane and worker nodes.
+Control plane nodes have access to the CA certificate and private key, therefore
+can easily refresh their own certificates. Worker nodes do not have access to
+the CA private key, so certificate refreshes need to happen in two steps.
+
+To accomodate this need, while also ensuring that the `k8s refresh-certs` can be
+straightforward to implement, two API calls are introduced, which clients should
+call in order.
+
+### `POST /refresh-certs/init`
+
+```go
+type RefreshCertificatesInitRequest struct {
+}
+
+type RefreshCertificatesInitResponse struct {
+  // Seed needs to be passed to `/refresh-certs/run` to ensure that CSR
+  // names are valid.
+  Seed int
+  // RequiredCertificateApprovals is a list of CSRs (kubectl get csr) that need
+  // to be manually approved when `/refresh-certs/run` is used. This list will
+  // be empty for control plane nodes.
+  RequiredCertificateApprovals []string
+}
+```
+
+`POST /refresh-certs/init` performs any necessary preparations for certificate
+refreshes on the current node.
+
+It returns a `seed` that must be passed to `POST /refresh-certs/run`. For worker
+nodes, it also returns a list of names of CertificateSigningRequest objects that
+will need to be approved and signed. This can be used by the CLI to print a
+helpful message and request that the certificates be signed.
+
+This method call is purely informative and does not have any side-effect on the
+cluster.
+
+### `POST /refresh-certs/run`
+
+```go
+type RefreshCertificatesRunRequest struct {
+  // Seed must match the seed from the received RefreshCertificatesInitResponse.
+  Seed int
+  // ExpirationDuration is the duration of the requested certificates.
+  ExpirationDuration time.Duration
+}
+
+type RefreshCertificatesRunResponse struct {
+  // ExpirationDuration is the duration of the new certificates.
+  ExpirationDuration time.Duration
```

Comment:
yes it can. someone might request for certificates for 10 years, we might not want to do that. instead, we will give them the maximum TTL we want, then include it here

Reply:
ack